In [24]:
%matplotlib inline
import pandas as pd
import numpy as np
import matplotlib as mpl
import matplotlib.pyplot as plt
import matplotlib.colors as colors
from mpl_toolkits.axes_grid1 import make_axes_locatable
# import seaborn as sns

from pandas import set_option
set_option("display.max_rows", 10)
pd.options.mode.chained_assignment = None

filename = 'facies_vectors.csv'
training_data = pd.read_csv(filename)
cesium_dataset = training_data.drop(['Formation', 'Well Name','Facies','NM_M','RELPOS','ILD_log10','DeltaPHI','PHIND','PE'], axis=1)
cesium_dataset.to_csv('cesium_dataset.csv', index=False, header=False)
training_data


Facies Formation        Well Name   Depth      GR  ILD_log10  DeltaPHI  \
0          3     A1 SH        SHRIMPLIN  2793.0  77.450      0.664     9.900   
1          3     A1 SH        SHRIMPLIN  2793.5  78.260      0.661    14.200   
2          3     A1 SH        SHRIMPLIN  2794.0  79.050      0.658    14.800   
3          3     A1 SH        SHRIMPLIN  2794.5  86.100      0.655    13.900   
4          3     A1 SH        SHRIMPLIN  2795.0  74.580      0.647    13.500   
...      ...       ...              ...     ...     ...        ...       ...   
4144       5      C LM  CHURCHMAN BIBLE  3120.5  46.719      0.947     1.828   
4145       5      C LM  CHURCHMAN BIBLE  3121.0  44.563      0.953     2.241   
4146       5      C LM  CHURCHMAN BIBLE  3121.5  49.719      0.964     2.925   
4147       5      C LM  CHURCHMAN BIBLE  3122.0  51.469      0.965     3.083   
4148       5      C LM  CHURCHMAN BIBLE  3122.5  50.031      0.970     2.609   

       PHIND     PE  NM_M  RELPOS  
0     11.915  4.600     1   1.000  
1     12.565  4.100     1   0.979  
2     13.050  3.600     1   0.957  
3     13.115  3.500     1   0.936  
4     13.300  3.400     1   0.915  
...      ...    ...   ...     ...  
4144   7.254  3.617     2   0.685  
4145   8.013  3.344     2   0.677  
4146   8.013  3.190     2   0.669  
4147   7.708  3.152     2   0.661  
4148   6.668  3.295     2   0.653  

[4149 rows x 11 columns]

In [2]:
training_data['Well Name'] = training_data['Well Name'].astype('category')
training_data['Formation'] = training_data['Formation'].astype('category')
training_data['Well Name'].unique()

[SHRIMPLIN, ALEXANDER D, SHANKLE, LUKE G U, KIMZEY A, CROSS H CATTLE, NOLAN, Recruit F9, NEWBY, CHURCHMAN BIBLE]
Categories (10, object): [SHRIMPLIN, ALEXANDER D, SHANKLE, LUKE G U, ..., NOLAN, Recruit F9, NEWBY, CHURCHMAN BIBLE]

In [3]:
# 1=sandstone  2=c_siltstone   3=f_siltstone 
# 4=marine_silt_shale 5=mudstone 6=wackestone 7=dolomite
# 8=packstone 9=bafflestone
facies_colors = ['#F4D03F', '#F5B041','#DC7633','#6E2C00',
       '#1B4F72','#2E86C1', '#AED6F1', '#A569BD', '#196F3D']

facies_labels = ['SS', 'CSiS', 'FSiS', 'SiSh', 'MS',
                 'WS', 'D','PS', 'BS']
#facies_color_map is a dictionary that maps facies labels
#to their respective colors
facies_color_map = {}
for ind, label in enumerate(facies_labels):
    facies_color_map[label] = facies_colors[ind]

def label_facies(row, labels):
    return labels[ row['Facies'] -1]
    
training_data.loc[:,'FaciesLabels'] = training_data.apply(lambda row: label_facies(row, facies_labels), axis=1)
training_data.describe()


/Users/lorenzoperozzi/anaconda3/lib/python3.5/site-packages/numpy/lib/function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


Facies        Depth           GR    ILD_log10     DeltaPHI  \
count  4149.000000  4149.000000  4149.000000  4149.000000  4149.000000   
mean      4.503254  2906.867438    64.933985     0.659566     4.402484   
std       2.474324   133.300164    30.302530     0.252703     5.274947   
min       1.000000  2573.500000    10.149000    -0.025949   -21.832000   
25%       2.000000  2821.500000    44.730000     0.498000     1.600000   
50%       4.000000  2932.500000    64.990000     0.639000     4.300000   
75%       6.000000  3007.000000    79.438000     0.822000     7.500000   
max       9.000000  3138.000000   361.150000     1.800000    19.312000   

             PHIND           PE         NM_M       RELPOS  
count  4149.000000  3232.000000  4149.000000  4149.000000  
mean     13.201066     3.725014     1.518438     0.521852  
std       7.132846     0.896152     0.499720     0.286644  
min       0.550000     0.200000     1.000000     0.000000  
25%       8.500000          NaN     1.000000     0.277000  
50%      12.020000          NaN     2.000000     0.528000  
75%      16.050000          NaN     2.000000     0.769000  
max      84.400000     8.094000     2.000000     1.000000

### Data exploration

In [4]:
# training_data.info()
print('There are {0:2.1f}% NaN in the photoelectric variable'.format(training_data['PE'].isnull().sum()/len(training_data['PE'])*100))

There are 22.1% NaN in the photoelectric variable


One solution to deal with NaN will be to dropp rows where there is NaN in PE variables (as Brandon Hall made in his tutorial). 
However, as 22.1% of data is a pretty good amount of data. Scikit-learn offers a builtin function to replace NaN with mean, median or most-frequent value. 
For this excercise we propose we prefer to replace NaN with a "clearly out-of-range" value ( like -99999). This allow ensemble classifier, that are based on decision-trees, to split on the criteria "known values vs unknown values". 

In [67]:
training_data['PE'].fillna(-99999, inplace=True)
training_data

# PE_mask = training_data['PE'].notnull().values
# training_data = training_data[PE_mask]
# training_data.describe()

Facies Formation        Well Name   Depth      GR  ILD_log10  DeltaPHI  \
0          3     A1 SH        SHRIMPLIN  2793.0  77.450      0.664     9.900   
1          3     A1 SH        SHRIMPLIN  2793.5  78.260      0.661    14.200   
2          3     A1 SH        SHRIMPLIN  2794.0  79.050      0.658    14.800   
3          3     A1 SH        SHRIMPLIN  2794.5  86.100      0.655    13.900   
4          3     A1 SH        SHRIMPLIN  2795.0  74.580      0.647    13.500   
...      ...       ...              ...     ...     ...        ...       ...   
4144       5      C LM  CHURCHMAN BIBLE  3120.5  46.719      0.947     1.828   
4145       5      C LM  CHURCHMAN BIBLE  3121.0  44.563      0.953     2.241   
4146       5      C LM  CHURCHMAN BIBLE  3121.5  49.719      0.964     2.925   
4147       5      C LM  CHURCHMAN BIBLE  3122.0  51.469      0.965     3.083   
4148       5      C LM  CHURCHMAN BIBLE  3122.5  50.031      0.970     2.609   

       PHIND     PE  NM_M  RELPOS  
0     11.915  4.600     1   1.000  
1     12.565  4.100     1   0.979  
2     13.050  3.600     1   0.957  
3     13.115  3.500     1   0.936  
4     13.300  3.400     1   0.915  
...      ...    ...   ...     ...  
4144   7.254  3.617     2   0.685  
4145   8.013  3.344     2   0.677  
4146   8.013  3.190     2   0.669  
4147   7.708  3.152     2   0.661  
4148   6.668  3.295     2   0.653  

[4149 rows x 11 columns]

Feauturize using cesium

In [91]:
F = training_data.as_matrix(columns=['Facies'])
F = training_data['Facies'].values
D = training_data.as_matrix(columns=['Depth'])
D = training_data['Depth'].values
M = training_data.as_matrix(columns=['GR'])
M = training_data['GR'].values

dict = {'Facies': F, 'Depth': D, 'GR': M}


array([3, 3, 3, ..., 5, 5, 5])

In [96]:
# import xarray as xr
# training_data_xr = xr.Dataset.from_dataframe(training_data)


from cesium import featurize
features_to_use = ['amplitude',
                   'percent_beyond_1_std',
                   'maximum',
                   'max_slope',
                   'median',
                   'median_absolute_deviation',
                   'percent_close_to_median',
                   'minimum',
                   'skew',
                   'std',
                   'weighted_average']
fset_cesium = featurize.featurize_time_series(times=D,
                                              values=M,
                                              errors=np.zeros(len(training_data['GR'])),
                                              features_to_use=features_to_use,
                                              targets=F)
print(fset_cesium)

ValueError: Wrong number of items passed 4149, placement implies 1

## Test set
Remove a single well to use as a blind test later.

In [6]:
blind = training_data[training_data['Well Name'] == 'NEWBY']
training_data = training_data[training_data['Well Name'] != 'NEWBY']
training_data['Well Name'].unique()
blind

Facies Formation Well Name   Depth     GR  ILD_log10  DeltaPHI  PHIND  \
3282       3     A1 SH     NEWBY  2826.0  76.34      0.719       7.8  11.00   
3283       3     A1 SH     NEWBY  2826.5  83.74      0.688       9.7  12.55   
3284       3     A1 SH     NEWBY  2827.0  83.19      0.664      10.1  11.95   
3285       3     A1 SH     NEWBY  2827.5  80.44      0.648      10.1  11.15   
3286       3     A1 SH     NEWBY  2828.0  75.42      0.648       9.3  11.45   
...      ...       ...       ...     ...    ...        ...       ...    ...   
3740       6      C LM     NEWBY  3055.0  66.94      0.838       4.0   8.00   
3741       6      C LM     NEWBY  3055.5  54.06      0.823       1.9   5.45   
3742       6      C LM     NEWBY  3056.0  47.87      0.797       0.7   4.85   
3743       6      C LM     NEWBY  3056.5  49.34      0.763       2.3   4.85   
3744       8      C LM     NEWBY  3057.0  59.88      0.721       3.2   7.20   

       PE  NM_M  RELPOS FaciesLabels  
3282  3.7     1   1.000         FSiS  
3283  3.4     1   0.977         FSiS  
3284  3.4     1   0.953         FSiS  
3285  3.4     1   0.930         FSiS  
3286  3.3     1   0.907         FSiS  
...   ...   ...     ...          ...  
3740  4.2     2   0.292           WS  
3741  4.3     2   0.281           WS  
3742  4.4     2   0.270           WS  
3743  4.1     2   0.258           WS  
3744  3.9     2   0.247           PS  

[463 rows x 12 columns]

## Split into training and test set

Now we extract just the feature variables we need to perform the classification.  The predictor variables are the five wireline values and two geologic constraining variables. We also get a vector of the facies labels that correspond to each feature vector.

In [7]:
X_train = training_data.drop(['Formation', 'Well Name', 'Depth','Facies','FaciesLabels'], axis=1).values
y_train = training_data['Facies'].values
X_test = blind.drop(['Formation', 'Well Name', 'Depth','Facies','FaciesLabels'], axis=1).values
y_test = blind['Facies'].values


print(X_train.shape)
print(X_test.shape)
print(y_train.shape)
print(y_test.shape)
# feature_vectors = training_data.drop(['Formation', 'Well Name', 'Depth','Facies','FaciesLabels'], axis=1)
# feature_vectors.describe()

(3686, 7)
(463, 7)
(3686,)
(463,)


In [9]:
from sklearn import cross_validation
from sklearn import tree
from sklearn import svm
from sklearn import ensemble
from sklearn import neighbors
from sklearn import linear_model
from sklearn import metrics
from sklearn import preprocessing

from sklearn.metrics import confusion_matrix
from classification_utilities import display_cm, display_adj_cm

def accuracy(conf):
    total_correct = 0.
    nb_classes = conf.shape[0]
    for i in np.arange(0,nb_classes):
        total_correct += conf[i][i]
    acc = total_correct/sum(sum(conf))
    return acc

adjacent_facies = np.array([[1], [0,2], [1], [4], [3,5], [4,6,7], [5,7], [5,6,8], [6,7]])

def accuracy_adjacent(conf, adjacent_facies):
    nb_classes = conf.shape[0]
    total_correct = 0.
    for i in np.arange(0,nb_classes):
        total_correct += conf[i][i]
        for j in adjacent_facies[i]:
            total_correct += conf[i][j]
    return total_correct / sum(sum(conf))

/Users/lorenzoperozzi/anaconda3/lib/python3.5/site-packages/sklearn/cross_validation.py:44: DeprecationWarning: This module was deprecated in version 0.18 in favor of the model_selection module into which all the refactored classes and functions are moved. Also note that the interface of the new CV iterators are different from that of this module. This module will be removed in 0.20.
  "This module will be removed in 0.20.", DeprecationWarning)


## Comparing classifier

In [12]:
classifiers = [
    svm.SVC( C=10, gamma = 1),
    ensemble.RandomForestClassifier(),
    ensemble.GradientBoostingClassifier(),
    ensemble.BaggingClassifier()]

The goal of ensemble methods is to combine the predictions of several base estimators built with a given learning algorithm in order to improve generalizability / robustness over a single estimator. Two families of ensemble methods are usually distinguished:
In averaging methods, the driving principle is to build several estimators independently and then to average their predictions. On average, the combined estimator is usually better than any of the single base estimator because its variance is reduced. Examples: Bagging methods, Forests of randomized trees, ...
By contrast, in boosting methods, base estimators are built sequentially and one tries to reduce the bias of the combined estimator. The motivation is to combine several weak models to produce a powerful ensemble. Examples: AdaBoost, Gradient Tree Boosting, ...

In [17]:
# Logging for Visual Comparison
log_cols=["Classifier", "Accuracy", "Precision", "Recall", "F1-score"]
log = pd.DataFrame(columns=log_cols)

for clf in classifiers:
    clf.fit(X_train, y_train)
    name = clf.__class__.__name__
    
    print("="*30)
    print(name)
    
    print('****Results****\n')
    predicted_labels = clf.predict(X_test)
    
#     accuracy = metrics.accuracy_score(y_test, train_predictions)
#     print("Accuracy: {:.4%}".format(accuracy))
    conf = confusion_matrix(y_test, predicted_labels)

    print('Optimized facies classification accuracy = %.2f' % accuracy(conf))
    print('Optimized adjacent facies classification accuracy = %.2f\n' % accuracy_adjacent(conf, adjacent_facies))
    
    display_adj_cm(conf, facies_labels, adjacent_facies,display_metrics=True, hide_zeros=True)

#     precision = metrics.precision_score(y_test, train_predictions, average='micro')
#     print("Precision: {:.4%}".format(precision))
    
#     recall = metrics.recall_score(y_test, train_predictions, average='micro')
#     print("Recall: {:.4%}".format(recall))
    
#     f1 = metrics.f1_score(y_test, train_predictions, average='micro')
#     print("F1 score: {:.4%}".format(f1))
       
#     log_entry = pd.DataFrame([[name, accuracy*100, precision*100, recall*100, f1*100]], columns=log_cols)
#     log = log.append(log_entry)
    
print("="*30)

SVC
****Results****

Optimized facies classification accuracy = 0.28
Optimized adjacent facies classification accuracy = 0.60

     Pred    SS  CSiS  FSiS  SiSh    MS    WS     D    PS    BS Total
     True
       SS                                                           0
     CSiS          97                                   1          98
     FSiS     2          76                 1           1          80
     SiSh          34    14     6           2     1     1          58
       MS          20                 7                 1          28
       WS          34     1     1          60                      96
        D          11                 1           4                16
       PS          25     2           1                28          56
       BS          30                 1                            31

Precision  0.00  0.39  0.82  0.86  0.70  0.95  0.80  0.88  0.00  0.70
   Recall  0.00  0.99  0.95  0.10  0.25  0.62  0.25  0.50  0.00  0.60
       F1  0.00  0.56 

In [ ]:
# from sklearn.metrics import confusion_matrix
# from classification_utilities import display_cm, display_adj_cm

# conf = confusion_matrix(y_test, train_predictions)
# display_cm(conf, facies_labels, hide_zeros=True)
display_adj_cm(conf, facies_labels, adjacent_facies, 
           display_metrics=True, hide_zeros=True)

In [ ]:
# conf = confusion_matrix(y_test, train_predictions)
# display_cm(conf, facies_labels, hide_zeros=True)
conf

In [ ]:
ax = sns.stripplot(x="Well Name", y="Facies", hue='Facies', data=training_data, jitter=False,palette="Set2")

In [ ]:
training_data['Facies']

In [ ]:
mask = training_data['PE'].isnull().values
training_data_test = training_data[mask]

In [ ]:
training_data_test['Well Name'].unique()

In [ ]:
def make_facies_log_plot(logs, facies_colors):
#     plt.style.use('seaborn-whitegrid')

    #make sure logs are sorted by depth
    logs = logs.sort_values(by='Depth')
    cmap_facies = colors.ListedColormap(
            facies_colors[0:len(facies_colors)], 'indexed')
    
    ztop=logs.Depth.min(); zbot=logs.Depth.max()
    
    cluster=np.repeat(np.expand_dims(logs['Facies'].values,1), 100, 1)
    
    f, ax = plt.subplots(nrows=1, ncols=6, figsize=(8, 12))
    ax[0].plot(logs.GR, logs.Depth, '-g')
    ax[1].plot(logs.ILD_log10, logs.Depth, '-')
    ax[2].plot(logs.DeltaPHI, logs.Depth, '-', color='0.5')
    ax[3].plot(logs.PHIND, logs.Depth, '-', color='r')
    ax[4].plot(logs.PE, logs.Depth, '-', color='black')
    im=ax[5].imshow(cluster, interpolation='none', aspect='auto',
                    cmap=cmap_facies,vmin=1,vmax=9)
    
    divider = make_axes_locatable(ax[5])
    cax = divider.append_axes("right", size="20%", pad=0.05)
    cbar=plt.colorbar(im, cax=cax)
    cbar.set_label((17*' ').join([' SS ', 'CSiS', 'FSiS', 
                                'SiSh', ' MS ', ' WS ', ' D  ', 
                                ' PS ', ' BS ']))
    cbar.set_ticks(range(0,1)); cbar.set_ticklabels('')
    
    for i in range(len(ax)-1):
        ax[i].set_ylim(ztop,zbot)
        ax[i].invert_yaxis()
        ax[i].grid()
        ax[i].locator_params(axis='x', nbins=3)
    
    ax[0].set_xlabel("GR")
    ax[0].set_xlim(logs.GR.min(),logs.GR.max())
    ax[1].set_xlabel("ILD_log10")
    ax[1].set_xlim(logs.ILD_log10.min(),logs.ILD_log10.max())
    ax[2].set_xlabel("DeltaPHI")
    ax[2].set_xlim(logs.DeltaPHI.min(),logs.DeltaPHI.max())
    ax[3].set_xlabel("PHIND")
    ax[3].set_xlim(logs.PHIND.min(),logs.PHIND.max())
    ax[4].set_xlabel("PE")
    ax[4].set_xlim(logs.PE.min(),logs.PE.max())
    ax[5].set_xlabel('Facies')
    
    ax[1].set_yticklabels([]); ax[2].set_yticklabels([]); ax[3].set_yticklabels([])
    ax[4].set_yticklabels([]); ax[5].set_yticklabels([])
    ax[5].set_xticklabels([])
    f.suptitle('Well: %s'%logs.iloc[0]['Well Name'], fontsize=14,y=0.94)

In [ ]:
facies_colors = ['#F4D03F', '#F5B041','#DC7633','#6E2C00',
       '#1B4F72','#2E86C1', '#AED6F1', '#A569BD', '#196F3D']

In [ ]:
make_facies_log_plot(
    training_data[training_data['Well Name']=='Recruit F9'],
    facies_colors)